### PR TITLE
[Feature] Support K-Net

### PR DESCRIPTION
**Pull request by MengzhangLI**
Re-implementing K-Net.

Original paper: https://arxiv.org/abs/2106.14855.
Original repo:https://github.com/ZwwWayne/K-Net.

### Results on ADE20K

| Method          | Backbone | Crop Size | Lr schd | Mem (GB) | Inf time (fps) | mIoU  | mIoU [from original repo](https://github.com/ZwwWayne/K-Net)
| --------------- | -------- | --------- | ------- | -------- | -------------- | ----------------- | :-:|
| KNet + FCN      | R-50-D8  | 512x512   | 80000   | 7.01     | 19.24          | **43.60** | 43.30 |
| KNet + PSPNet   | R-50-D8  | 512x512   | 80000   | 6.98     | 20.04          | **44.18** | 43.90 |
| KNet + DeepLabV3| R-50-D8  | 512x512   | 80000   | 7.42     | 12.10          | **45.06** | 44.60 |
| KNet + UperNet  | R-50-D8  | 512x512   | 80000   | 7.34     | 17.11          | 43.45 | **43.60** |
| KNet + UperNet  | Swin-T   | 512x512   | 80000   | 7.57     | 15.56          | **45.84** | 45.40 |
| KNet + UperNet  | Swin-L   | 512x512   | 80000   | 13.5     | 8.29           | **52.05** | 52.0 |
| KNet + UperNet  | Swin-L   | 640x640   | 80000   | 13.54    | 8.29           | 52.21 | **52.7** |

### CODE REVIEW

**Comment by MengzhangLI** [comment]
After refactoring, the model can directly used for inference and result is the same, which means the key is still the same with old one.

image

**Comment by MengzhangLI** [comment]
Add ep3,4,5.

**Comment by MengzhangLI** [comment]
Could anyone @ZwwWayne for code review? Thx in advance.

File: mmseg/models/decode_heads/uper_head.py [hunk 1]
```diff
@@ -84,9 +84,12 @@ def psp_forward(self, inputs):
 
         return output
 
-    def forward_feature(self, inputs):
-        """Feature map before `self.cls_seg` and learnable semantic kernels can
-        be both output for kernel updation."""
+    def _forward_feature(self, inputs):
```
Comment (ZwwWayne) [inline mmseg/models/decode_heads/uper_head.py:87]: Is change forward_feature to _forward_feature necessary? It might cause BC-breaking

**Comment by MeowZheng** [comment]
`_forward_feature` was added most recently
ref: https://github.com/open-mmlab/mmsegmentation/pull/1299

File: mmseg/models/decode_heads/knet.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        features = self.fc_norm(features)
+        features = self.activation(features)
+
+        return features
+
+
+@HEADS.register_module()
+class KernelUpdateHead(nn.Module):
+
+    def __init__(self,
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/knet.py:115]: Need to add docstring

File: mmseg/models/decode_heads/uper_head.py (hunk 1 above)
Comment (MengzhangLI) [reply mmseg/models/decode_heads/uper_head.py:87]: We have fixed it by: https://github.com/open-mmlab/mmsegmentation/pull/1299.

**Review by MeowZheng — COMMENTED** [review]
might add readme file and metafile

File: mmseg/models/decode_heads/knet_head.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+        activate_out: Whether add activation after gate mechanism.
+            Default: False.
+        norm_cfg (dict | None): Config of norm layers.
+            Default: dict(type='LN').
+        act_cfg (dict): Config of activation layers.
+            Default: dict(type='ReLU').
+
+    Returns:
+        Tensor: The output tensor of shape
+        (self.num_classes(KernelUpdateHead) * self.in_channels(KernelUpdateHead)/ in_channels, kernel size * kernel size, in_channels). # noqa
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/knet_head.py:38]: `__init__` returns None, so might remove returns in docstring

File: mmseg/models/decode_heads/knet_head.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+        self.activation = build_activation_layer(act_cfg)
+
+        self.fc_layer = nn.Linear(self.feat_channels, self.out_channels, 1)
+        self.fc_norm = build_norm_layer(norm_cfg, self.out_channels)[1]
+
+    def forward(self, update_feature, input_feature):
+        # `input_feature` is dynamic kernels for updation with shape:
+        # (N, num_classes, conv_kernel_size * conv_kernel_size, channels),
+        # it would be reshaped to `update_feature` whose last dimension
+        # shape is `self.in_channels`.
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/knet_head.py:89]: might rewrite docstring in google style

File: mmseg/models/decode_heads/knet_head.py [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+                     type='DynamicConv',
+                     in_channels=256,
+                     feat_channels=64,
+                     out_channels=256,
+                     act_cfg=dict(type='ReLU', inplace=True),
+                     norm_cfg=dict(type='LN')).
+
+    Returns:
+        Tensor: The mask prediction of shape (N, num_classes, H, W).
+        Tensor: The dynamic kernels of shape (N, num_classes, channels, K, K).
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/knet_head.py:185]: `__init__` doesn't have returns

File: mmseg/models/decode_heads/knet_head.py [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+
+        Args:
+            x (Tensor): Feature map from FPN with shape
+                (batch_size, feature_dimensions, H , W).
+            proposal_feat (Tensor): Intermediate feature get from
+                diihead in last stage, has shape
+                (batch_size, num_proposals, feature_dimensions)
+            mask_preds (Tensor): mask prediction from the former stage in shape
+                (batch_size, num_proposals, H, W).
+        """
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/knet_head.py:290]: might add some description for returns

File: mmseg/models/decode_heads/knet_head.py [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+        num_stages (int): The number of stages (kernel update heads)
+            in IterativeDecodeHead. Default: 3.
+        kernel_generate_head:(dict): Config of kernel generate head which
+            generate mask predictions, dynamic kernels and class predictions
+            for next kernel update heads.
+        kernel_update_head (dict): Config of kernel update head which refine
+            dynamic kernels and class predictions iteratively.
+
+    Returns:
+        Tensor: The output tensor of shape (N, out_channels, H, W).
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/knet_head.py:405]: migth remove

File: mmseg/models/decode_heads/knet_head.py [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+        Args:
+            update_feature (torch.Tensor): Feature map assembled from
+                each group. It would be reshaped with last dimension
+                shape: `self.in_channels`.
+            input_feature (torch.Tensor): Intermediate feature
+                with shape: (N, num_classes, conv_kernel_size**2, channels).
+
+        Returns:
+            Tensor: The output tensor of shape
+            (self.num_classes(KernelUpdateHead) * self.in_channels(KernelUpdateHead)/ in_channels, kernel size * kernel size, in_channels). # noqa
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/knet_head.py:93]: Tensor:  The dynamic kernel with shape ()?
is there any short description for the shape of the dynamic kernel? @MengzhangLI @ZwwWayne

File: mmseg/models/decode_heads/knet_head.py [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+        return stage_segs[-1]
+
+    def losses(self, seg_logit, seg_label):
+        losses = dict()
+        for i, logit in enumerate(seg_logit):
+            loss = self.kernel_generate_head.losses(logit, seg_label)
+            for k, v in loss.items():
+                losses[f'{k}.s{i}'] = v
+
+        return losses
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/knet_head.py:453]: might add ut  for loss

**Review by MeowZheng — APPROVED** [review]
